### DISCUSSION TITLE
Unloader positioning

### DISCUSSION THREAD

**Discussion by coatsy35** [Courseplay/Courseplay_FS22#3183] [General]
@pvaiko Noticed that the unloader drives very close to the forager with wide maize headers on. Pictures of CP driving tractor very close to a forager with a wide header (does the same on all wide giants maize headers), then a picture of it driving further away with a narrow header. Surely it should drive outside the width of the header just like IRL and like combines do? I have noticed a number of times the unloader then tries to drive forward over the maize header resulting in a crash when it's full - not managed to capture it with a log yet. But you can see why it happens as the unloader is behind the header, when it should surely be outside of the header. 

Also the unloader in the bottom shot is not behind the forager as I assume it hasn't hit the trigger for too much fruit to be either side to pull in behind - doesn't matter to me as I never play with crop destruction on, but for those that do, perhaps a slight tweak in the calculation is in order? so it goes behind as soon as there is crop either side?

![fsScreen_2024_03_30_16_27_32](https://github.com/Courseplay/Courseplay_FS22/assets/122838698/f334c527-07a4-4df8-84d8-1d3731f1fb76)

![fsScreen_2024_03_30_21_06_47](https://github.com/Courseplay/Courseplay_FS22/assets/122838698/7577e78a-26d8-4c78-918a-762f7955400d)

**Comment by Tensuko** [comment]
> Noticed that the unloader drives very close to the forager with wide maize headers on.

intended.

> then a picture of it driving further away with a narrow header.

intended, if you notice something, the position does not depend on the header...

> I have noticed a number of times the unloader then tries to drive forward over the maize header resulting in a crash when it's full

Maybe we should keep on the 2nd sensor untill we are far away enough.

> But you can see why it happens as the unloader is behind the header, when it should surely be outside of the header.

Thats not the reason.

> Also the unloader in the bottom shot is not behind the forager as I assume it hasn't hit the trigger for too much fruit to be either side to pull in behind

If there is enough space (yea, its a bit close) the unloader won't drive behind the chopper.

**Comment by pvaiko** [comment]
> Surely it should drive outside the width of the header just like IRL and like combines do?

Plenty of videos where it is not, like here https://www.facebook.com/watch/?v=307181452044381

The lateral distance is based on the vehicle width and tries to keep at least one meter between them. (The width is as defined by the shop, the area the game needs to place a vehicle after you buy it, so it is wider than the actual width).

Which side we are driving, is based on the amount of fruit on each side of the harvester, for that we use a Giants function which the Giant helper invokes to decide which way to turn, so in your case there's less fruit on the left, so we don't stay behind the harvester, which we would normally. No idea how to get into the situation on your screenshot, normally, when using our patterns you have nearly equal amount of fruit on both sides so we stay in the middle.

Forward position is tractor reference node 1 m behind the harvester's reference node, except when the tractor's proximity controller sees the header, it will stay back more.

**Reply by rohne83** [reply]
Haven't had the opportunity yet to test the forage unloader, but just few annotations from me. To me it looks like you both are right ;)
IRL the default unloader position is as far to the front that the pipe can have an angle of ~90 degrees (think that's what coatsy's trying to depict with his attached videos). Independent from header width - but of course only if there's no fruit on that side of the harvester. If there's fruit to the side, then the unloaders stay in position as you depicted with that facebook vid, or exactly behind the harvester if the header is smaller. Reason why the pipe of the second harvester there is not in 90 degrees angle seems just to be the small distance to the 1st harvester. With larger distance we would see both there - ~45degrees pipe for 1st and 90degrees pipe for 2nd harvester.
Big question for me is, how is the pipe side decided i.e. on grass fields? Especially when thinking of trailers with overblow protection (as i.e. the HW80s on modhub have).

**Comment by coatsy35** [comment]
> Plenty of videos where it is not, like here https://www.facebook.com/watch/?v=307181452044381
> 
> The lateral distance is based on the vehicle width and tries to keep at least one meter between them. (The width is as defined by the shop, the area the game needs to place a vehicle after you buy it, so it is wider than the actual width).

Yes when your in the middle of the field you drive not quite behind the unit (I have done plenty of chopping IRL driving both the chopper and the unloader). When you have not got crop on one side you would drive out and away from the header as far as practical, never intentionally being close to it at all - especially true for students who are inexperienced 🤣

Germany
https://youtu.be/NwPXuK31B2I?si=sErOJD2kQrGGCZij&t=178

UK
https://www.youtube.com/watch?v=y9bXTnmCyhA

USA
https://youtu.be/ikpExS10rvI?si=BNbTEVUm2ayeYlWP&t=75

Good examples of how it's usually done when working at the side of the work after opening up a land 

> No idea how to get into the situation on your screenshot, normally, when using our patterns you have nearly equal amount of fruit on both sides so we stay in the middle.

That happens when you come in from an angled edge on a field where there is less fruit for a while on one side than the other (in my screenshot I had taken slice out of a field with another smaller machine, but it happens on angled parts of the field - as I said I don't mind it, but anyone who uses fruit destruction might, was just an observation. 

IT is fantastic work the team has done though, so good to see forager mode back 😎 precall second unloader one day ?...
